<a href="https://colab.research.google.com/github/dgg32/ml_pycaret/blob/main/pycaret_anomaly_detection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install pycaret

     |████████████████████████████████| 256kB 11.9MB/s 
     |████████████████████████████████| 1.8MB 18.3MB/s 
     |████████████████████████████████| 102kB 10.0MB/s 
     |████████████████████████████████| 276kB 46.8MB/s 
     |████████████████████████████████| 6.8MB 49.4MB/s 
     |████████████████████████████████| 266kB 36.9MB/s 
     |████████████████████████████████| 174kB 53.8MB/s 
     |████████████████████████████████| 1.6MB 39.6MB/s 
     |████████████████████████████████| 148.9MB 83kB/s 
     |████████████████████████████████| 66.3MB 60kB/s 
     |████████████████████████████████| 13.9MB 49.9MB/s 
     |████████████████████████████████| 2.1MB 44.3MB/s 
     |████████████████████████████████| 604kB 43.6MB/s 
     |████████████████████████████████| 71kB 9.8MB/s 
     |████████████████████████████████| 3.1MB 36.3MB/s 
     |████████████████████████████████| 71kB 9.7MB/s 
     |████████████████████████████████| 61kB 7.2MB/s 
     |████████████████████████████████| 81kB 8.8MB/s 


In [2]:
from pycaret.utils import enable_colab
enable_colab()

Colab mode enabled.


# What is Anomaly Detection?

Anomaly Detection is the task of identifying the rare items, events or observations which raise suspicions by differing significantly from the majority of the data. Typically the anomalous items will translate to some kind of problem such as bank fraud, a structural defect, medical problems or errors in a text. There are three broad categories of anomaly detection techniques exist:

    Unsupervised anomaly detection: Unsupervised anomaly detection techniques detect anomalies in an unlabeled test data set under the assumption that the majority of the instances in the dataset are normal by looking for instances that seem to fit least to the remainder of the data set.

    Supervised anomaly detection: This technique requires a dataset that has been labeled as "normal" and "abnormal" and involves training a classifier.

    Semi-supervised anomaly detection: This technique constructs a model representing normal behavior from a given normal training dataset, and then tests the likelihood of a test instance to be generated by the learnt model.

pycaret.anomaly module supports the unsupervised and supervised anomaly detection technique. In this tutorial we will only cover unsupervised anomaly detection technique.

# Overview of Anomaly Detection Module in PyCaret

PyCaret's anomaly detection module (pycaret.anomaly) is a an unsupervised machine learning module which performs the task of identifying rare items, events or observations which raise suspicions by differing significantly from the majority of the data.

PyCaret anomaly detection module provides several pre-processing features that can be configured when initializing the setup through setup() function. It has over 12 algorithms and few plots to analyze the results of anomaly detection. PyCaret's anomaly detection module also implements a unique function tune_model() that allows you to tune the hyperparameters of anomaly detection model to optimize the supervised learning objective such as AUC for classification or R2 for regression.

# Dataset for the Tutorial

For this tutorial we will use a dataset from UCI called Mice Protein Expression. The dataset consists of the expression levels of 77 proteins/protein modifications that produced detectable signals in the nuclear fraction of cortex. The dataset contains a total of 1080 measurements per protein. Each measurement can be considered as an independent sample/mouse. Click Here to read more about the dataset.

In [3]:
from pycaret.datasets import get_data
dataset = get_data("mice")

MouseID  DYRK1A_N   ITSN1_N    BDNF_N  ...  Genotype  Treatment  Behavior   class
0   309_1  0.503644  0.747193  0.430175  ...   Control  Memantine       C/S  c-CS-m
1   309_2  0.514617  0.689064  0.411770  ...   Control  Memantine       C/S  c-CS-m
2   309_3  0.509183  0.730247  0.418309  ...   Control  Memantine       C/S  c-CS-m
3   309_4  0.442107  0.617076  0.358626  ...   Control  Memantine       C/S  c-CS-m
4   309_5  0.434940  0.617430  0.358802  ...   Control  Memantine       C/S  c-CS-m

[5 rows x 82 columns]

In [4]:
dataset.shape

(1080, 82)

In order to demonstrate the predict_model() function on unseen data, a sample of 5% (54 samples) are taken out from original dataset to be used for predictions at the end of experiment. This should not be confused with train/test split. This particular split is performed to simulate real life scenario. Another way to think about this is that these 54 samples are not available at the time when this experiment was performed.

In [5]:
data = dataset.sample(frac=0.95, random_state=786)
data_unseen = dataset.drop(data.index)

data.reset_index(drop=True, inplace=True)
data_unseen.reset_index(drop=True, inplace=True)

print (f"Data for modeling {data.shape}")
print (f"Unseen Data for predictions: {data_unseen.shape}")

Data for modeling (1026, 82)
Unseen Data for predictions: (54, 82)


# Setting up Environment in PyCaret

setup() function initializes the environment in PyCaret and creates the transformation pipeline to prepare the data for modeling and deployment. setup() must be called before executing any other function in PyCaret. It takes only one mandatory parameter: pandas dataframe. All other parameters are optional and are used to customize pre-processing pipeline (we will see them in later tutorials).

When setup() is executed, PyCaret's inference algorithm will automatically infer the data types for all features based on certain properties. Although, most of the times the data type is inferred correctly but it's not always the case. Therefore, after setup() is executed, PyCaret displays a table containing features and their inferred data types. At which stage, you can inspect and press enter to continue if all data types are correctly inferred or type quit to end the experiment. Identifying data types correctly is of fundamental importance in PyCaret as it automatically performs few pre-processing tasks which are imperative to perform any machine learning experiment. These pre-processing tasks are performed differently for each data type. As such, it is very important that data types are correctly configured.

In later tutorials we will learn how to overwrite PyCaret's inferred data types using numeric_features and categorical_features parameter in setup()

In [6]:
from pycaret.anomaly import *

exp_ano101 = setup(data, normalize=True, ignore_features=["MouseID"], session_id=123)

Description                 Value
0                               session_id                   123
1                            Original Data            (1026, 82)
2                           Missing Values                  True
3                         Numeric Features                    77
4                     Categorical Features                     4
5                         Ordinal Features                 False
6                High Cardinality Features                 False
7                  High Cardinality Method                  None
8                         Transformed Data            (1026, 91)
9                                 CPU Jobs                    -1
10                                 Use GPU                 False
11                          Log Experiment                 False
12                         Experiment Name  anomaly-default-name
13                                     USI                  7afd
14                         Imputation Type                simple
15          Iterative Imputation Iteration                  None
16                         Numeric Imputer                  mean
17      Iterative Imputation Numeric Model                  None
18                     Categorical Imputer                  mode
19  Iterative Imputation Categorical Model                  None
20           Unknown Categoricals Handling        least_frequent
21                               Normalize                  True
22                        Normalize Method                zscore
23                          Transformation                 False
24                   Transformation Method                  None
25                                     PCA                 False
26                              PCA Method                  None
27                          PCA Components                  None
28                     Ignore Low Variance                 False
29                     Combine Rare Levels                 False
30                    Rare Level Threshold                  None
31                         Numeric Binning                 False
32                         Remove Outliers                 False
33                      Outliers Threshold                  None
34                Remove Multicollinearity                 False
35             Multicollinearity Threshold                  None
36                              Clustering                 False
37                    Clustering Iteration                  None
38                     Polynomial Features                 False
39                       Polynomial Degree                  None
40                    Trignometry Features                 False
41                    Polynomial Threshold                  None
42                          Group Features                 False
43                       Feature Selection                 False
44            Features Selection Threshold                  None
45                     Feature Interaction                 False
46                           Feature Ratio                 False
47                   Interaction Threshold                  None



Once the setup is succesfully executed it prints the information grid that contains few important information. Much of the information is related to pre-processing pipeline which is constructed when setup() is executed. Much of these features are out of scope for the purpose of this tutorial. However, few important things to note at this stage are:

    session_id : A pseudo-random number distributed as a seed in all functions for later reproducibility. If no session_id is passed, a random number is automatically generated that is distributed to all functions. In this experiment session_id is set as 123 for later reproducibility.

    Missing Values : When there are missing values in original data it will show as True. Notice that Missing Values in the information grid above is True as the data contains missing values which are automatically imputed using mean for numeric features and constant for categorical features. The method of imputation can be changed using numeric_imputation and categorical_imputation parameter in setup().

    Original Data : Displays the original shape of dataset. In this experiment (1026, 82) means 1026 samples and 82 features.

    Transformed Data : Displays the shape of transformed dataset. Notice that the shape of original dataset (1026, 82) is transformed into (1026, 91). The number of features has increased due to encoding of categorical features in the dataset.

    Numeric Features : Number of features inferred as numeric. In this dataset, 77 out of 82 features are inferred as numeric.

    Categorical Features : Number of features inferred as categorical. In this dataset, 5 out of 82 features are inferred as categorical. Also notice, we have ignored one categorical feature i.e. MouseID using ignore_feature parameter.

Notice that how few tasks such as missing value imputation and categorical encoding that are imperative to perform modeling are automatically handled. Most of the other parameters in setup() are optional and used for customizing pre-processing pipeline. These parameters are out of scope for this tutorial but as you progress to intermediate and expert level, we will cover them in much detail.


# Create a Model

Creating an anomaly detection model in PyCaret is simple and similar to how you would have created a model in supervised modules of PyCaret. The anomaly detection model is created using create_model() function which takes one mandatory parameter i.e. name of the model as a string. This function returns a trained model object. See the example below:

In [7]:
iforest = create_model("iforest")

In [8]:
print (iforest)

IForest(behaviour='new', bootstrap=False, contamination=0.05,
    max_features=1.0, max_samples='auto', n_estimators=100, n_jobs=-1,
    random_state=123, verbose=0)


We have created Isolation Forest model using create_model(). Notice the contamination parameter is set 0.05 which is the default value when you do not pass fraction parameter in create_model(). fraction parameter determines the proportion of outliers in the dataset. In below example, we will create One Class Support Vector Machine model with 0.025 fraction.

In [9]:
svm = create_model("svm", fraction=0.025)

In [10]:
print (svm)

OCSVM(cache_size=200, coef0=0.0, contamination=0.025, degree=3, gamma='auto',
   kernel='rbf', max_iter=-1, nu=0.5, shrinking=True, tol=0.001,
   verbose=False)




Just by replacing iforest with svm inside create_model() we have now created OCSVM anomaly detection model. There are 12 models available ready-to-use in pycaret.anomaly module. To see the complete list, please see docstring or use models function.


In [11]:
models()

Name                    Reference
ID                                                                       
abod            Angle-base Outlier Detection        pyod.models.abod.ABOD
cluster       Clustering-Based Local Outlier      pyod.models.cblof.CBLOF
cof         Connectivity-Based Local Outlier          pyod.models.cof.COF
iforest                     Isolation Forest  pyod.models.iforest.IForest
histogram  Histogram-based Outlier Detection        pyod.models.hbos.HBOS
knn             K-Nearest Neighbors Detector          pyod.models.knn.KNN
lof                     Local Outlier Factor          pyod.models.lof.LOF
svm                   One-class SVM detector      pyod.models.ocsvm.OCSVM
pca             Principal Component Analysis          pyod.models.pca.PCA
mcd           Minimum Covariance Determinant          pyod.models.mcd.MCD
sod               Subspace Outlier Detection          pyod.models.sod.SOD
sos             Stochastic Outlier Selection          pyod.models.sos.SOS

# Assign a Model

Now that we have created a model, we would like to assign the anomaly labels to our dataset (1080 samples) to analyze the results. We will achieve this by using assign_model() function. See an example below:

In [12]:
iforest_results = assign_model(iforest)
iforest_results.head()

MouseID  DYRK1A_N   ITSN1_N  ...   class  Anomaly  Anomaly_Score
0  3501_12  0.344930  0.626194  ...  t-SC-m        0      -0.014462
1   3520_5  0.630001  0.839187  ...  c-CS-m        0      -0.070193
2  3414_13  0.555122  0.726229  ...  t-CS-m        0      -0.070143
3   3488_8  0.275849  0.430764  ...  c-SC-s        0      -0.080521
4   3501_7  0.304788  0.617299  ...  t-SC-m        0      -0.064749

[5 rows x 84 columns]

Notice that two columns Label and Score are added towards the end. 0 stands for inliers and 1 for outliers/anomalies. Score is the values computed by the algorithm. Outliers are assigned with larger anomaly scores. Notice that iforest_results also includes MouseID feature that we have dropped during setup(). It wasn't used for the model and is only appended to the dataset when you use assign_model(). In the next section we will see how to analyze the results of anomaly detection using plot_model().

# plot a model

plot_model() function can be used to analyze the anomaly detection model over different aspects. This function takes a trained model object and returns a plot. See the examples below:

In [15]:
plot_model(iforest)

In [14]:
plot_model(iforest, plot="umap")

# Predict on Unseen Data

predict_model() function is used to assign anomaly labels on the new unseen dataset. We will now use our iforest model to predict the data stored in data_unseen. This was created in the beginning of the experiment and it contains 54 new samples that were not exposed to PyCaret before.

In [17]:
unseen_predictions = predict_model(iforest, data=data_unseen)
unseen_predictions.head()

DYRK1A_N   ITSN1_N    BDNF_N  ...   class  Anomaly  Anomaly_Score
0  0.447506  0.628176  0.367388  ...  c-CS-m        0      -0.077131
1  0.704633  0.802537  0.350110  ...  c-CS-m        0      -0.060165
2  0.505093  0.695549  0.376029  ...  c-CS-m        0      -0.052132
3  0.429133  0.563175  0.258429  ...  c-CS-m        0      -0.091610
4  0.373648  0.471165  0.257909  ...  c-CS-m        1       0.037436

[5 rows x 83 columns]

Label column indicates the outlier (1 = outlier, 0 = inlier). Score is the values computed by the algorithm. Outliers are assigned with larger anomaly scores. You can also use predict_model() function to label the training data. See example below:

In [18]:
data_predictions = predict_model(iforest, data=data)
data_predictions.head()

DYRK1A_N   ITSN1_N    BDNF_N  ...   class  Anomaly  Anomaly_Score
0  0.344930  0.626194  0.383583  ...  t-SC-m        0      -0.014462
1  0.630001  0.839187  0.357777  ...  c-CS-m        0      -0.070193
2  0.555122  0.726229  0.278319  ...  t-CS-m        0      -0.070143
3  0.275849  0.430764  0.285166  ...  c-SC-s        0      -0.080521
4  0.304788  0.617299  0.335164  ...  t-SC-m        0      -0.064749

[5 rows x 83 columns]

# Saving the Model

We have now finished the experiment by using our iforest model to predict outlier labels on unseen data. This brings us to the end of our experiment but one question is still to be asked. What happens when you have more new data to predict? Do you have to go through the entire experiment again? The answer is No, you don't need to rerun the entire experiment and reconstruct the pipeline to generate predictions on new data. PyCaret's inbuilt function save_model() allows you to save the model along with entire transformation pipeline for later use.

In [19]:
save_model(iforest, "Final IForest Model 18Nov2020")

Transformation Pipeline and Model Succesfully Saved


(Pipeline(memory=None,
          steps=[('dtypes',
                  DataTypes_Auto_infer(categorical_features=[],
                                       display_types=True,
                                       features_todrop=['MouseID'],
                                       id_columns=[],
                                       ml_usecase='classification',
                                       numerical_features=[],
                                       target='UNSUPERVISED_DUMMY_TARGET',
                                       time_features=[])),
                 ('imputer',
                  Simple_Imputer(categorical_strategy='most frequent',
                                 fill_value_categorical=None,
                                 fill_value_n...
                 ('fix_perfect', 'passthrough'),
                 ('clean_names', Clean_Colum_Names()),
                 ('feature_select', 'passthrough'), ('fix_multi', 'passthrough'),
                 ('dfs', 'passthrough'), ('

# Loading the Saved Model

To load a saved model on a future date in the same or different environment, we would use the PyCaret's load_model() function and then easily apply the saved model on new unseen data for prediction

In [20]:
saved_iforest = load_model("Final IForest Model 18Nov2020")

Transformation Pipeline and Model Successfully Loaded


Once the model is loaded in the environment, you can simply use it to predict on any new data using the same predict_model() function . Below we have applied the loaded model to predict the same data_unseen that we have used in section 10 above.

In [21]:
new_prediction = predict_model(saved_iforest, data = data_unseen)
new_prediction.head()

DYRK1A_N   ITSN1_N    BDNF_N  ...   class  Anomaly  Anomaly_Score
0  0.447506  0.628176  0.367388  ...  c-CS-m        0      -0.077131
1  0.704633  0.802537  0.350110  ...  c-CS-m        0      -0.060165
2  0.505093  0.695549  0.376029  ...  c-CS-m        0      -0.052132
3  0.429133  0.563175  0.258429  ...  c-CS-m        0      -0.091610
4  0.373648  0.471165  0.257909  ...  c-CS-m        1       0.037436

[5 rows x 83 columns]

Notice that results of unseen_predictions and new_prediction are identical.